src/Servers/Kestrel/Core/src/Systemd/KestrelServerOptionsSystemdExtensions.cs

```diff
@@ -36,7 +37,17 @@ public static KestrelServerOptions UseSystemd(this KestrelServerOptions options,
         {
             if (string.Equals(Process.GetCurrentProcess().Id.ToString(CultureInfo.InvariantCulture), Environment.GetEnvironmentVariable(ListenPidEnvVar), StringComparison.Ordinal))
             {
-                options.ListenHandle(SdListenFdsStart, configure);
+                if (!byte.TryParse(Environment.GetEnvironmentVariable(ListenFdsEnvVar), NumberStyles.None, NumberFormatInfo.InvariantInfo, out var listenFds))
+                {
+                    listenFds = 1;
+                }
+
+                for (ulong handle = SdListenFdsStart, lastHandle = SdListenFdsStart + listenFds;
+                    handle <= lastHandle;
+                    ++handle)
```

Comment:
If `listenFds` has a default value of 1, this would listen to 2 handles (3 and 4). I think you want `<` not `<=`. Also, as a nit, I think the above should be one line. Maybe

```
                for (var handle = SdListenFdsStart; handle < SdListenFdsStart + listenFds; handle++)
```

Reply:
Yes, I didn't test this yet, I'll describe the reasons below.

Reply:
Fixed this, but please take a look at my comment below on the docker-based test.